src/qibolab/_core/instruments/keysight/sweep.py

```diff
@@ -0,0 +1,110 @@
+"""Utils for sweeper management."""
+
+from collections import defaultdict
+
+import keysight.qcs as qcs  # pylint: disable=E0401
+
+from qibolab._core.identifier import ChannelId
+from qibolab._core.pulses import PulseId
+from qibolab._core.sweeper import ParallelSweepers, Parameter
+
+NS_TO_S = 1e-9
+
+__all__ = ["process_sweepers"]
+
+
+def process_sweepers(
+    sweepers: list[ParallelSweepers], probe_channel_ids: set[ChannelId]
+):
+    """Processes Qibocal sweepers into QCS sweepers. Currently nested hardware
+    sweepers are not supported, so they will default to software sweeping.
+
+    Arguments:
+        sweepers (list[ParallelSweepers]): Array of array of sweepers.
+        probe_channel_ids (list[int]): Array of channel IDs for probe channels.
+
+    Returns:
+        hardware_sweepers (list[tuple[list[qcs.Array], list[qcs.Scalar]]]): Array of hardware-based QCS sweepers.
+        software_sweepers (list[tuple[list[qcs.Array], list[qcs.Scalar]]]): Array of software-based QCS sweepers.
+        sweeper_swaps_required (list[tuple[int, int]]): Array of corresponding axes to be swapped in results.
+        sweeper_channel_map (dict[ChannelId, qcs.Scalar]): Map of channel ID to frequency to be swept.
+        sweeper_pulse_map (defaultdict[PulseId, dict[str, qcs.Scalar]]): Map of pulse ID to map of parameter
+        to be swept and corresponding QCS variable.
+    """
+    hardware_sweepers: list[tuple[list[qcs.Array], list[qcs.Scalar]]] = []
+    software_sweepers: list[tuple[list[qcs.Array], list[qcs.Scalar]]] = []
+    hw_sweeper_order = []
+    sw_sweeper_order = []
+
+    # Mapper for pulses that are controlled by a sweeper and the parameter to be swept
+    sweeper_pulse_map: defaultdict[PulseId, dict[str, qcs.Scalar]] = defaultdict(dict)
+    # Mapper for channels with frequency controlled by a sweeper
+    sweeper_channel_map: dict[ChannelId, qcs.Scalar] = {}
+
+    for idx, parallel_sweeper in enumerate(sweepers):
+        sweep_values: list[qcs.Array] = []
+        sweep_variables: list[qcs.Variable] = []
+        # Currently nested hardware sweeping is not supported
+        hardware_sweeping = len(hardware_sweepers) == 0
+
+        for idx2, sweeper in enumerate(parallel_sweeper):
+            qcs_variable = qcs.Scalar(
+                name=f"V{idx}_{idx2}", value=sweeper.values[0], dtype=float
+            )
+
+            if sweeper.parameter is Parameter.frequency:
+                sweeper_channel_map.update(
+                    {channel_id: qcs_variable for channel_id in sweeper.channels}
+                )
```

Comment:
Just to clarify: Qibolab itself is not planning to become an FPGA design, but the idea is to have a high-level language to be compiled to the individual ones. But at the same time, to keep it as close as possible to the existing languages to minimize the conversion (i.e. compilation) efforts.

If then you're willing to contribute to the language design, and you may want to use it as the basis to design an FPGA-based processor (at which point, it would require essentially no conversion), you're definitely more than welcome :P